### PR TITLE
TNT: add explode_center flag

### PR DESCRIPTION
**Pull request by tenplus1**
Add 'explode_middle' flag which when false explodes as normal and when true runs on_blast on centre node as well as dropping items.

EDIT by paramat: addresses #1825

### CODE REVIEW

**Comment by paramat** [comment]
@sofar

File: mods/tnt/init.lua [hunk 1]
```diff
@@ -385,9 +389,11 @@ function tnt.boom(pos, def)
 	local meta = minetest.get_meta(pos)
 	local owner = meta:get_string("owner")
 	minetest.sound_play("tnt_explode", {pos = pos, gain = 1.5, max_hear_distance = 2*64})
-	minetest.set_node(pos, {name = "tnt:boom"})
+	if not def.explode_centre then
+		minetest.set_node(pos, {name = "tnt:boom"})
+	end
 	local drops, radius = tnt_explode(pos, def.radius, def.ignore_protection,
-			def.ignore_on_blast, owner)
+			def.ignore_on_blast, owner, def.explosion_centre)
```
Comment (sfan5) [inline mods/tnt/init.lua]: mismatch with doc

**Comment by sfan5** [comment]
`explode_center`*

**Comment by paramat** [comment]
MT uses USA spelling 'center' in the code, although the UK spelling is present in a few places in the docs. So please use 'center'.

**Comment by sofar** [comment]
I'm not sure this solves the problem. What if I make a sheep that explodes by calling `tnt.boom()`, and it happens to explode while inside a `default:grass_1` node? I'm not registering a new tnt node, so nowhere am I passing this extra flag to the TNT API.

**Comment by tenplus1** [comment]
@sofar - for exploding sheep you would call the tnt.boom function like so:
```
tnt.boom(pos, {
	radius = radius,
	damage_radius = radius,
	explode_centre = true, -- explodes even without TNT
	sound = sound, -- if pull #1823 gets included
})

**Comment by sofar** [comment]
Right, I see I misread the patch.

**Comment by DS-Minetest** [comment]
Shouldn't `"tnt:boom"` also be set with `explode_centre = true` but later?

**Comment by tenplus1** [comment]
@DS-Minetest - I don't think so, if explode_centre isn't set then it'll remove middle block (usually tnt) and explode from there.

**Comment by DS-Minetest** [comment]
@tenplus1 Yes but it doesn't set `"tnt:boom"` at all if  `explode_centre == true`. And the `"tnt:boom"` is for the appearance, isn't it?

**Comment by tenplus1** [comment]
The tnt:boom node is counted in the overall blast for normal TNT and provides light when exploding, without it we still have a tnt_boom particle effect that is lit even when tnt:boom isn't added.

**Comment by DS-Minetest** [comment]
It's not consistent that there's no light emitting node with `explode_centre` enabled. Use the `tnt:boom` node in both cases or don't use it at all.

**Comment by tenplus1** [comment]
@DS-Minetest - the tnt.boom function when used with explode_centre flag will be used for mobs and other entities exploding like cannonballs etc. so to have it drop the centre node is important when it hits something...  To replace this centre node with tnt:boom wont help with drops and/or running the centre node's on_boom functions.

File: mods/tnt/init.lua [hunk 2]
```diff
@@ -385,9 +390,11 @@ function tnt.boom(pos, def)
...
```
Comment (DS-Minetest) [inline mods/tnt/init.lua]: Add here
```lua
if def.explode_centre then
	minetest.set_node(pos, {name = "tnt:boom"})
end
```
for consistency.

**Comment by DS-Minetest** [comment]
Ok, I understand now. :+1:

**Comment by sofar** [comment]
:+1: on all 3 patches but please change `centre` to `center`.

**Comment by paramat** [comment]
Yes for MT code spelling consistency.

**Comment by tenplus1** [comment]
https://github.com/minetest/minetest/search?utf8=%E2%9C%93&q=centre&type= <-- we use the word 'centre' many times without code and documentation so my current spelling will stay as-is.

**Comment by jastevenson303** [comment]
And will you also change words like meter (metre) and color (colour)? Where do you draw the line?

**Comment by tenplus1** [comment]
@jastevenson303 - exactly, the code should stand for itself and not spellings used.

**Comment by paramat** [comment]
'centre' may be used in code, but 'centre' is used more often, and docs don't count because they can be easily changed without breakage. I prefer the English spelling too, but code consistency is more important.

**Comment by jastevenson303** [comment]
I don't understand what you mean.  Are you proposing changing all instances of 'color' in Minetest to 'colour'?

**Comment by tenplus1** [comment]
"You're traveling through another dimension, a dimension not only of sight and sound but of mind. A journey into a wondrous land whose boundaries are that of imagination. That's the signpost up ahead - your next stop, the Twilight Zone!"

**Comment by paramat** [comment]
This got me interested so i checked, no use of 'centre' in code anywhere in engine or MTGame https://github.com/minetest/minetest/issues/6238

**Comment by tenplus1** [comment]
Has this pull been put on hold because of regional spelling ?

**Comment by paramat** [comment]
LOL middle :] I love how you're getting around this.
'Center' is a much better word to use, used in most places in Minetest code when a centre is referred to.
Also see https://github.com/minetest/minetest_game/blob/master/mods/tnt/init.lua#L30 it's already used in the TNT mod. So it's obvious we should use 'center'.
Don't worry, i'm not irritated, you can use whatever word you want, today i have just been explaining the reasoning.
So :-1: for 'middle', sorry. Would you allow us to alter this on merge or should someone else make a new PR?

**Comment by tenplus1** [comment]
@paramat - was worth a try...

File: game_api.txt [hunk 3]
```diff
@@ -309,7 +309,8 @@ TNT API
 ^ Create an explosion.
 
 * `position` The center of explosion.
-* `definition` The TNT definion as passed to `tnt.register`
+* `definition` The TNT definion as passed to `tnt.register` with the following addition:
+ * `explode_center` false by default which removes TNT node on blast, when true will explode centre node.
```
Comment (paramat) [inline game_api.txt]: Missed one at end of line.

**Comment by paramat** [comment]
Tried to merge but conflicts with your other TNT PR.

**Comment by paramat** [comment]
Tried to merge, even with 3-way merge, but did not apply.

**Comment by tenplus1** [comment]
@paramat - it seems to show "All checks have passed" so cannot figure where rebase is needed ?

**Comment by paramat** [comment]
> it seems to show "All checks have passed"

Yes this happens sometimes, the actual changed lines do not conflict but nearby ones do, in this situation a 3-way merge often works but this time it didn't.

**Pull request merged**